### DISCUSSION TITLE
How does this compare to PAPI?

### DISCUSSION THREAD

**Discussion by sanmai-NL** [LLNL/variorum#392] [Q&A] [answered]
So [PAPI](https://icl.utk.edu/papi/#overview) is about general performance ... But also supports energy measurement. What's the motivation for developing Variorum independently?

**Comment by tpatki** [comment, marked as answer]
Hi @sanmai-NL : 
PAPI does not allow for control dials (except limited Intel RAPL power capping support). Variorum supports dials such as setting frequency, socket or memory power caps across all the platforms/architectures that Variorum is implemented for.  

For Intel and AMD systems, Variorum also supports these dials through user-space (https://github.com/LLNL/msr-safe), which PAPI does not. PAPI also does not have JSON based interfaces to export data easily to higher-level system software, such as resource managers (Flux) or profiling and runtime tools (Caliper/Kokkos). Variorum's integration with Flux and other tools allows for end-to-end dynamic power management at all levels for power management. 

Please take a look at the HPC PowerStack and our ECP Tutorial to learn more about dynamic power management.

- Details on PowerStack: https://variorum.readthedocs.io/en/latest/PowerStack.html 

- Module 2 of the ECP Tutorial Series talks about dynamic power management in detail: https://variorum.readthedocs.io/en/latest/Publications.html

**Reply by sanmai-NL** [reply]
PAPI (now?) supports [`libsmr`](https://github.com/icl-utk-edu/papi/wiki/PAPI-Overview#list-of-components).

**Reply by rountree** [reply]
libmsr support in PAPI happened in 2015.  Heike flew out to Livermore to discuss integration with myself, Tapasya, and the other authors of libmsr.  

We mothballed libmsr five years ago and replaced it with Variorum.

At some point the PAPI folks might want to update their code to use Variorum, but there doesn't appear to be much crossover between the two communities.  People who are satisfied with the underlying Linux perf interface use PAPI, and people who need functionality beyond that either use Variorum or program against the msr-safe API directly.
